Suppress clang warnings

minoki opened the pull request:
This PR suppresses some clang warnings described below:
1. `GetClassInfo()` in `wxDECLARE_ABSTRACT_CLASS` [-Winconsistent-missing-override]
   This is a virtual function defined in the macro. Adding `override` to the function can potentially break user code (as commented in Trac), so I concluded that using compiler pragmas is the best fix (or workaroud) here.
   See also [Trac #17017](http://trac.wxwidgets.org/ticket/17017).
2. Some virtual functions defined in `DECLARE_HELP_PROVISION()` [-Winconsistent-missing-override]
   Similar situation to 1.
3. "Expression with side effects ..." in `WX_DECLARE_ANY_VALUE_TYPE` [-Wpotentially-evaluated-expression]
   This one is reported and a fix was attempted in [Trac #16968](http://trac.wxwidgets.org/ticket/16968), but on my environment (clang 3.6.2 and 3.7.0), that fix was not working. So I'm just using pragmas to silence the compiler.

I added a pair of macros to silence these warnings, named `wxCLANG_WARNING_SUPPRESS_IF_VALID` and `wxCLANG_WARNING_RESTORE_IF_VALID`. They work like `wxGCC_WARNING_SUPPRESS` and `wxGCC_WARNING_RESTORE`, but only emit the pragma when the compiler recognizes the warning. So there will be no complaint of unknown-warning-name from GCC or older clang (~3.5).

minoki committed:
Suppress clang warning [-Winconsistent-missing-override] for GetClassInfo() in wxDECLARE_ABSTRACT_CLASS.

minoki committed:
Suppress clang warning [-Winconsistent-missing-override] in DECLARE_HELP_PROVISION().

minoki committed:
Add macros to conditionally suppress clang warnings.

The added macros wxCLANG_WARNING_{SUPPRESS,RESTORE} are similar to
wxGCC_WARNING_{SUPPRESS,RESTORE}, but the `wxCLANG_' version will expand to
compiler pragmas only when the warning name is known to the compiler.

They use clang's __has_warning feature.

vadz commented:
I might be missing something obvious but how exactly could adding `wxOVERRIDE` to `wxDECLARE_ABSTRACT_CLASS` break any user code? I just don't see it...

minoki commented:
If we add `wxOVERRIDE` to `wxDECLARE_ABSTRACT_CLASS`, the compilation of a root class such as `wxObject` will fail.  Although I expect there to be few examples of such class outside wxWidgets.

Another effect will be that there can be extra warnings for the user code that doesn't adopt `override`.  Consider the following code.  If we just add `override` to `GetClassInfo()`, the compiler will issue a warning for `Bar::greet`.
```c++
class wxClassInfo;

class Foo
{
    virtual void greet() {}
    virtual wxClassInfo *GetClassInfo() { return nullptr; }
};

// A simplified version of `wxDECLARE_ABSTRACT_CLASS` for illustration

#if USE_PRAGMA
// The version I propose
// No warnings whether or not the user-code adopts `override`.
#define wxDECLARE_ABSTRACT_CLASS(name) \
    _Pragma("clang diagnostic push") \
    _Pragma("clang diagnostic ignored \"-Winconsistent-missing-override\"") \
    virtual wxClassInfo *GetClassInfo() { return nullptr; } \
    _Pragma("clang diagnostic pop")

#elif USE_OVERRIDE
// A version that add `override` keyword
#define wxDECLARE_ABSTRACT_CLASS(name) \
    virtual wxClassInfo *GetClassInfo() override { return nullptr; }

#else
// Current wx version
#define wxDECLARE_ABSTRACT_CLASS(name) \
    virtual wxClassInfo *GetClassInfo() { return nullptr; }
#endif

/*
 * A class that doesn't adopt `override` keyword.
 *
 * If compiled with `USE_OVERRIDE`, a warning for greet() is issued.
 * No warnings with `USE_PRAGMA`.
 */
class Bar : public Foo
{
    virtual void greet() {}
    wxDECLARE_ABSTRACT_CLASS(Bar);
};

/*
 * A class that does use `override` keyword.
 *
 * If compiled without `USE_OVERRIDE`, a warning for `GetClassInfo()` is issued.
 * No warnings with `USE_PRAGMA`.
 */
class Bar2 : public Foo
{
    virtual void greet() override {}
    wxDECLARE_ABSTRACT_CLASS(Bar2);
};
```

minoki commented:
I've updated the branch.

- The warning in `include/wx/any.h` was finally fixed in master (see [Trac #16968](http://trac.wxwidgets.org/ticket/16968)). So no need to fix it here.
- Removed the suffix `_IF_VALID` as [you suggested](https://github.com/minoki/wxWidgets/commit/53a34f758a29dab94d8fddb04e6983d6b9af1564#commitcomment-13577658).

vadz commented:
Thank you for your explanations and the updated patch, I'll apply it soon.
